File: examples/dictionary_compression.c
/*
 * Copyright (c) 2016-present, Yann Collet, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under both the BSD-style license (found in the
 * LICENSE file in the root directory of this source tree) and the GPLv2 (found
 * in the COPYING file in the root directory of this source tree).
 * You may select, at your option, one of the above-listed licenses.
 */
#include <stdlib.h>    // malloc, exit
#include <stdio.h>     // printf
#include <string.h>    // strerror
#include <errno.h>     // errno
#include <sys/stat.h>  // stat
#include <zstd.h>      // presumes zstd library is installed
#include "utils.h"

/* createDict() :
   `dictFileName` is supposed to have been created using `zstd --train` */
static ZSTD_CDict* createCDict_orDie(const char* dictFileName, int cLevel)
{
    size_t dictSize;
    printf("loading dictionary %s \n", dictFileName);
    void* const dictBuffer = mallocAndLoadFile_orDie(dictFileName, &dictSize);
    ZSTD_CDict* const cdict = ZSTD_createCDict(dictBuffer, dictSize, cLevel);
    if (!cdict) {
        fprintf(stderr, "ZSTD_createCDict error \n");
        exit(7);
    }
    free(dictBuffer);
    return cdict;
}


static void compress(const char* fname, const char* oname, const ZSTD_CDict* cdict)
{
    size_t fSize;
    void* const fBuff = mallocAndLoadFile_orDie(fname, &fSize);
    size_t const cBuffSize = ZSTD_compressBound(fSize);
    void* const cBuff = malloc_orDie(cBuffSize);

    ZSTD_CCtx* const cctx = ZSTD_createCCtx();
    if (cctx==NULL) { fprintf(stderr, "ZSTD_createCCtx() error \n"); exit(10); }
    size_t const cSize = ZSTD_compress_usingCDict(cctx, cBuff, cBuffSize, fBuff, fSize, cdict);
    if (ZSTD_isError(cSize)) {
        fprintf(stderr, "error compressing %s : %s \n", fname, ZSTD_getErrorName(cSize));
        exit(7);
    }

    saveFile_orDie(oname, cBuff, cSize);

    /* success */
    printf("%25s : %6u -> %7u - %s \n", fname, (unsigned)fSize, (unsigned)cSize, oname);

    ZSTD_freeCCtx(cctx);   /* never fails */
    free(fBuff);
    free(cBuff);
}


static char* createOutFilename_orDie(const char* filename)
{
    size_t const inL = strlen(filename);
    size_t const outL = inL + 5;
    void* outSpace = malloc_orDie(outL);
    memset(outSpace, 0, outL);
    strcat(outSpace, filename);
    strcat(outSpace, ".zst");
    return (char*)outSpace;
}

int main(int argc, const char** argv)
{
    const char* const exeName = argv[0];
    int const cLevel = 3;

    if (argc<3) {
        fprintf(stderr, "wrong arguments\n");
        fprintf(stderr, "usage:\n");
        fprintf(stderr, "%s [FILES] dictionary\n", exeName);
        return 1;
    }

    /* load dictionary only once */
    const char* const dictName = argv[argc-1];
    ZSTD_CDict* const dictPtr = createCDict_orDie(dictName, cLevel);

    int u;
    for (u=1; u<argc-1; u++) {
        const char* inFilename = argv[u];
        char* const outFilename = createOutFilename_orDie(inFilename);
        compress(inFilename, outFilename, dictPtr);
        free(outFilename);
    }

    ZSTD_freeCDict(dictPtr);
    printf("All %u files compressed. \n", argc-2);
    return 0;
}

File: examples/dictionary_compression.zig
const std = @import("std");
const zstd = @import("zstd");
const common = @import("common.zig");

pub fn main() !void {
    const allocator = std.heap.c_allocator;
    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);

    if (args.len < 3) {
        std.debug.print("wrong arguments\nusage:\n{s} [FILES] dictionary\n", .{args[0]});
        return;
    }

    const dict_filename = args[args.len - 1];
    std.debug.print("loading dictionary {s} \n", .{dict_filename});

    const dict_buffer = try common.readFile(allocator, dict_filename);
    defer allocator.free(dict_buffer);

    const cLevel = 3;
    const cdict = zstd.c.ZSTD_createCDict(dict_buffer.ptr, dict_buffer.len, cLevel);
    if (cdict == null) {
        std.debug.print("zstd.eateCDict() failed!\n", .{});
        return;
    }
    defer _ = zstd.c.ZSTD_freeCDict(cdict);

    const cctx = zstd.c.ZSTD_createCCtx();
    if (cctx == null) {
        std.debug.print("zstd.eateCCtx() failed!\n", .{});
        return;
    }
    defer _ = zstd.c.ZSTD_freeCCtx(cctx);

    var i: usize = 1;
    while (i < args.len - 1) : (i += 1) {
        const input_filename = args[i];
        const input_data = try common.readFile(allocator, input_filename);
        defer allocator.free(input_data);

        const dest_size = zstd.c.ZSTD_compressBound(input_data.len);
        const dest_buffer = try allocator.alloc(u8, dest_size);
        defer allocator.free(dest_buffer);

        const cSize = zstd.c.ZSTD_compress_usingCDict(cctx, dest_buffer.ptr, dest_size, input_data.ptr, input_data.len, cdict);
        if (zstd.c.ZSTD_isError(cSize) != 0) {
            std.debug.print("error compressing: {s}\n", .{zstd.c.ZSTD_getErrorName(cSize)});
            continue;
        }

        const out_filename = try common.createOutFilename(allocator, input_filename);
        defer allocator.free(out_filename);
        try common.writeFile(out_filename, dest_buffer[0..cSize]);

        std.debug.print("{s} : {d} -> {d} - {s}\n", .{
            input_filename,
            input_data.len,
            cSize,
            out_filename,
        });
    }

    std.debug.print("All {d} files compressed. \n", .{args.len - 2});
}



